### PR TITLE
Implement Auth.SSO

### PR DESCRIPTION
**Pull request by pbrisbin**
- REVERT ME: Configure for Stack

  I had trouble with the nix setup; this worked straight away. I can remove it
  before merge.

- Expose Auth.ConfigFile functions for re-use

  The way ConfigFile works right now, I couldn't draw the perfect seams to
  totally DRY what I'm doing; but this seems like a good 80/20.

- Implement Auth.SSO

  Huzzah!

### CODE REVIEW

File: lib/amazonka/amazonka.cabal
Comment (endgame) [inline lib/amazonka/amazonka.cabal:99]: Please add bounds here. From looking at the changelog I think `>= 0.9 && <0.11` should be safe. Then `cabal-fmt --inplace` the cabal file.

File: .gitignore
Comment (endgame) [inline .gitignore:55]: This is probably worth keeping, even once you remove your stack commit from the PR.

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:104]: This probably needs to use `configPathRelative`.

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:105]: From that botocore snippet I put on the linked issue, we might be able to guess the name of the config file if we can find the right URL to hash? (If we can't, then globbing is fine)

File: lib/amazonka/src/Amazonka/Auth.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth.hs:114]: I don't think this will be necessary because we should be doing all the SSO lookup when we handle config files in `fromFileEnv`. The CLI doesn't document that it can do SSO from environment variables, so we can assume that it won't and that we won't need a `fromSSOEnv` either.

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:9]: Is it actually writing uuids, or the sha of `start_url`?

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:95]: I'd like to see this pulled out somewhere and used to DRY up `Amazonka.Auth.STS`' `assumeRoleWithWebIdentity` support. I don't think it needs to be part of the public interface, but maybe `Amazonka.Auth.Internal.sendWithoutAuth` (new module) would be good?

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:31]:
```suggestion
import qualified Data.Text as Text
import qualified Data.Text.Encoding as Text
```

**Review by endgame — CHANGES_REQUESTED** [review]
Thanks for PRing this; I think it will be a great help to a lot of people. I've picked a couple of nits, but I think there's a more fundamental restructuring that needs to happen first.

The decision to try SSO as part of looking up a config profile should happen in `ConfigFile.hs`, where it should be treated as "just another auth method". This probably means extending the `ConfigProfile` type (the parsers for the different profiles are in the code immediately above; you will probably want to make a new constructor for SSO):

https://github.com/brendanhay/amazonka/blob/29e4888f847797533c968a855804990ecf8660dd/lib/amazonka/src/Amazonka/Auth/ConfigFile.hs#L216-L227

Then in this `case`-expression, check for your new SSO constructor and in that branch, go looking for cached JSON files. For each one you find, call into the functions you provide in `Amazonka.Auth.SSO`:

https://github.com/brendanhay/amazonka/blob/29e4888f847797533c968a855804990ecf8660dd/lib/amazonka/src/Amazonka/Auth/ConfigFile.hs#L116-L140

When implemented this way, it will get handled in the same way as all other config-file-based methods, which means this hack to set region will apply to it also:

https://github.com/brendanhay/amazonka/blob/29e4888f847797533c968a855804990ecf8660dd/lib/amazonka/src/Amazonka/Auth/ConfigFile.hs#L84-L86

Please also add an entry for this PR (including giving yourself credit!) to `lib/amazonka/CHANGELOG.md`; once we have the names of the new SSO functions nailed down the big table of auth methods will need a new entry too.

*****

> Attempting to use this when an expired `{uuid}.json` is present results in:
> 
> ```
> ServiceError (...snip...)
> ```
> 
> Is that acceptable, or should we catch it and emit something clearer?

I think it's acceptable for now; STS errors seem to be thrown in the same way. If we want to pull them into Auth errors in response to user complaints, that's fine but I don't think we have to hold up this PR to do it.

**Comment by pbrisbin** [comment]
Thank you for the detailed review! I misunderstood some things, but I think I understand better now.

I'll basically be adding support for seeing the `sso_*` fields in `ConfigFile` and, if present, that's where/when to go into the SSO path. This inverts things a little from what I have now, but is a better design (which you know, of course).

The `sso_start_url`, which I'll have handy at that point, then determines which file under `~/.aws/sso/cache` to look for/parse, which solves a number of uncertainties and feels far more reliable. This is a great catch, I was thrown off by the fact that my two Accounts are the same start-url in my use-case.

I'll be sure to format, bounds, and changelog once I get the new approach dialed in. I should be able to push another attempt tonight or tomorrow.

On the topic of formatting, would you be interested in my [Restyled](https://restyled.io) App on this project? It has support for ormolu and I could add cabal-fmt (and anything else really) relatively quickly.

File: lib/amazonka/src/Amazonka/Auth/ConfigFile.hs
Comment (pbrisbin) [inline lib/amazonka/src/Amazonka/Auth/ConfigFile.hs:145]: Open to ideas here. Passing the `startUrl` to `fromSSO` and having it call `configPathRelative` internally would make much more sense to me, but causes an import loop. This was the best I could come up with.

File: lib/amazonka/src/Amazonka/Auth/ConfigFile.hs
Comment (pbrisbin) [inline lib/amazonka/src/Amazonka/Auth/ConfigFile.hs:241]:
```suggestion
  | -- | Recognizes @sso_start_url@, @sso_region@, @sso_account_id@, and
```

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (pbrisbin) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:76]: Agreed this should be centralized. Is there is a reason it can't go with `send`/`sendEither`? I'm not sure it needs to be considered "internal" -- I can imagine users wanting to send un-authed requests themselves in some cases.

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (pbrisbin) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:82]: Need to match the following Python,

```py
hashlib.sha1(start_url.encode('utf-8')).hexdigest()
```

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:46]: As with the comment in `ConfigFile.hs`, maybe rename to `cachedTokenFilePath`?

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:62]: Would you like to fix up the service definition, or shall we proceed as-is?

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:83]: Can you please add a comment with that link to the `botocore` repo, and an explanation why we're doing this? Because so much of this behaviour is determined by reading source, it's better if we can make that easier for the devs who come after us. Example:

https://github.com/brendanhay/amazonka/blob/29e4888f847797533c968a855804990ecf8660dd/lib/amazonka-core/src/Amazonka/Request.hs#L182-L183

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:111]: I think deriving Functor is reasonable.

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:107]: Can you please hang this under a `where` clause in `roleCredentialsToAuthEnv`?

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:96]: Please use `mconcat` when accumulating text like this - it works out how long the string needs to be, and only has to allocate once.

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:82]:
```suggestion
  let sha1 = show . Crypto.hashSHA1 $ Text.encodeUtf8 startUrl
```

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:48]: Question: It's correct to not update region on the returned environment, right? This is because the SSO region and the region the library user wants might be different. Even though we respect `AWS_REGION` somewhat, we might also have the region we want in the input `Env' withAuth`.

If I've got things right, would you mind adding a comment here? This seems a bit subtle.

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:38]: Please re-export this function from `Amazonka.Auth`, and add a haddock.

**Review by endgame — CHANGES_REQUESTED** [review]
All right, this is looking much better, thanks for your efforts. There are a lot of comments in this review, but most of them should be easy to address.

File: lib/amazonka/src/Amazonka/Auth/ConfigFile.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/ConfigFile.hs:243]: Maybe we call this `AssumeRoleViaSSO` for consistency with the other constructors?

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (pbrisbin) [reply lib/amazonka/src/Amazonka/Auth/SSO.hs:31]: Yeah, originally I was OK with that for an internal type, it's actually "SSO.Cache" when you think about it, but the convention of open exports in this module damages that approach. I'll do `CachedAccessToken` though it is more than that: the fields are `{startUrl, region, accessToken, expires, expiresAt}`.

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (pbrisbin) [reply lib/amazonka/src/Amazonka/Auth/SSO.hs:107]: Sure, care to elaborate why?

At $work, we have a convention to always make something that _can_ be top-level top-level -- we find it helps with thinking about parts in isolation (vs lumping the complexity of this function into the other one as a single "thing"), results in more type signatures (being now non-optional), and will very often end up being re-used (my need to move `configFileRelative` to accomplish this PR muddies its diff when it wouldn't need to had this been followed).

Not pushing back at all, just curious to hear the other side.

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (pbrisbin) [reply lib/amazonka/src/Amazonka/Auth/SSO.hs:96]: TIL! Apparently, I have **a lot** of my own code to go and update now :grimacing:

File: lib/amazonka/src/Amazonka/Auth/ConfigFile.hs
Comment (pbrisbin) [reply lib/amazonka/src/Amazonka/Auth/ConfigFile.hs:145]: I see by your second comment you mean `cache`. Done.

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (pbrisbin) [reply lib/amazonka/src/Amazonka/Auth/SSO.hs:73]: It's up to you. I'd need some hand-holding, but I'm happy to do it if you'd like.

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (pbrisbin) [reply lib/amazonka/src/Amazonka/Auth/SSO.hs]: Renamed to `sendUnsigned` going to save the rest for later.

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (pbrisbin) [reply lib/amazonka/src/Amazonka/Auth/SSO.hs:59]: I added a comment on the use of `ssoRegion` that I think addresses this. Going to optimistically Resolve conversation.

**Comment by pbrisbin** [comment]
OK, I think everything is resolved with the exception of the service definition.

For the CHANGELOG, can you offer some guidance? I'm not sure if this is a "Major Change" or "Other Changes".

I'd also like to cleanup the git History before merge, to isolate the feature from other pre/re-factors -- unless you happen to have a "Squash" policy (or some other preference). Do you?

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [reply lib/amazonka/src/Amazonka/Auth/SSO.hs]: Agreed that it can be done separately. Created #758 - do you want to own this? If not, it's small enough that I can get it done.

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [reply lib/amazonka/src/Amazonka/Auth/SSO.hs:76]: OK, if you're willing to wait for this, then let's do it properly.

Comment (endgame) [reply lib/amazonka/src/Amazonka/Auth/SSO.hs:76]: #759 - I'll merge in a day or so in case anyone has comments, then regenerate.

File: lib/amazonka/src/Amazonka/Auth/ConfigFile.hs
Comment (endgame) [reply lib/amazonka/src/Amazonka/Auth/ConfigFile.hs]: Yeah I think it makes sense conceptually, even though it's not an `sts:AssumeRole` or `sts:AssumeRoleWithWebIdentity` call.

File: lib/amazonka/src/Amazonka/Auth/ConfigFile.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/ConfigFile.hs:182]: Nit: please rename this variable to `assumeRoleViaSSO` for consistency.

File: lib/amazonka/src/Amazonka/Auth/ConfigFile.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/ConfigFile.hs:242]: I appreciate your diligence maintaining haddock conventions.

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:97]:
```suggestion
-- [botocore](https://github.com/boto/botocore/blob/c02f3561f56085b8a3f98501d25b9857b916c10e/botocore/utils.py#L2596-L2597),
-- so that we find tokens produced by @aws sso login@.
```

**Review by endgame — CHANGES_REQUESTED** [review]
Thanks very much. One last round of nitpicking, otherwise I'm pretty happy with how this looks. I'm also happy with your CHANGELOG entries, so thanks for updating the big table.

Here's how I see the next few days going:

1. Leave #759 open for a day or two in case someone objects.
2. (Me) Merge #759 and regenerate service bindings.
3. (You) While waiting for #759, rebase/rearrange commits until you're happy. I usually merge rather than squash in this repository, so if you can make the history look nice that'd be cool. But I also wouldn't lose sleep over it.
4. (You) Rebase on top of `main` once the regenerated `amazonka-sso` is available, so you're not coding against an errant `Maybe`; also remove the temporary stack files.
5. (Me) Merge this PR.

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (pbrisbin) [reply lib/amazonka/src/Amazonka/Auth/SSO.hs:97]: Whoa, when did Haddock start supporting markdown for links? I had to look it up to confirm but you're totally right.

**Comment by pbrisbin** [comment]
OK! I think this is done from my end. I will,

- Wait for #759 
- Rebase after that's merged

Then you can Approve/Merge.

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:40]:
```suggestion
-- | Assume a role using an SSO Token.
```

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:54]:
```suggestion
  -- | Account ID
  Text ->
  -- | Role Name
  Text ->
```

File: lib/amazonka/src/Amazonka/Auth/SSO.hs
Comment (endgame) [inline lib/amazonka/src/Amazonka/Auth/SSO.hs:43]: This haddock is actually describing the behaviour of `fromConfigFile`. Would you mind rewriting to say something like:

```
The user must have previously called @aws sso login@, and pass in the path to the cached token file, along with SSO region, account ID and role name. ('Amazonka.Auth.ConfigFile.fromFilePath' understands the @sso_@ variables used by the official AWS CLI and will call 'fromSSO' for you.) This function uses 'fetchAuthInBackground'...
```

**Review by endgame — APPROVED** [review]
Great, now we wait for the new version of service definitions. I don't expect anyone to object so I'll do them over the next couple of days.

**Comment by endgame** [comment]
The situation with service definitions and the generator is more complex than I first anticipated. I'm going to merge this now, as removing the `Maybe` is an easy fix when I do a generator pass.

**Comment by endgame** [comment]
Thanks again for your work and patience.

**Comment by pbrisbin** [comment]
Thank you for shepherding this through, this'll make using our amazonka-based projects so much easier.

**Pull request merged**
